src/call/mod.rs

```diff
@@ -683,4 +690,43 @@ impl SinkBase {
         self.batch_f.take();
         Poll::Ready(Ok(()))
     }
+
+    #[inline]
+    fn poll_flush<C: ShareCallHolder>(
+        &mut self,
+        cx: &mut Context,
+        call: &mut C,
+    ) -> Poll<Result<()>> {
+        if self.batch_f.is_some() {
+            ready!(self.poll_ready(cx)?);
+        }
+        if self.buf_flags.is_some() {
+            self.start_send_buffer_message(self.buf_buffer_hint, call)?;
+            ready!(self.poll_ready(cx)?);
+        }
+        Poll::Ready(Ok(()))
+    }
+
+    #[inline]
+    fn start_send_buffer_message<C: ShareCallHolder>(
+        &mut self,
+        buffer_hint: bool,
+        call: &mut C,
+    ) -> Result<()> {
+        let mut flags = self.buf_flags.clone().unwrap();
+        flags = flags.buffer_hint(buffer_hint);
+        let write_f = call.call(|c| {
+            c.call
+                .start_send_message(&self.buf, flags.flags, self.send_metadata)
+        })?;
+        self.batch_f = Some(write_f);
```

Comment:
This can go wrong. Let's say user sends [(msg1, not_bufferred), (msg2, bufferred)], old implement will only keep msg2 in buffer, but new implement will keep both.

Reply:
This is the meaning of this PR, buffer the messages in a stream as much as possible